apps/src/util/useFetch.js

```diff
@@ -0,0 +1,93 @@
+import {useState, useEffect} from 'react';
+
+const baseFetchState = {
+  loading: false,
+  data: null,
+  error: null
+};
+
+/**
+ * React hook for making a fetch request to a resource that returns a JSON
+ * response.
+ *
+ * By default, fetch requests are configured with the following options:
+ * - credentials: 'same-origin'
+ * The default options can be overriden by the passed-in options.
+ *
+ * The 'loading' field of the return value indicates whether there is currently
+ * a request in flight. Note that this value will initially be false and then
+ * transition to true as a request is sent and internal state is updated
+ * asynchronously.
+ *
+ * If the fetch request is successful, the body is parsed as json and the
+ * resulting object is put in the 'data' field of the return value. If the
+ * request is not successful or the body could not be parsed as json, the
+ * 'error' field will contain an object representing the error.
+ *
+ * @param {string} url - URL of resource to fetch
+ * @param {RequestInit} options - options to pass to fetch
+ * @param {any[]} deps - array of values that fetch depends on; a new fetch
+ *    request will be sent if any of the values in the array change
+ * @returns {{loading: boolean, data: Object, error: Object}}
+ */
+export const useFetch = (url, options, deps) => {
+  const [fetchState, setFetchState] = useState(baseFetchState);
```

Comment:
I originally had it as three separate pieces of state but was worried about seeing inconsistent intermediate states (e.g. loading = true + data not null) due to `set*` being asynchronous.  I combined these into a single state object so that they would change atomically.

Another approach I saw in some libraries was to use useRef (so values would update synchronously) and "force updates" but I thought that was a bit hacky.

Thoughts?

Reply:
Oh yep, I didn't know about that actually! React will batch those calls at the top level, but setState in async calls do not get batched. This makes sense then!